pkg/cmd/create/create_devpod.go

```diff
@@ -277,7 +277,8 @@ func (o *CreateDevPodOptions) Run() error {
 		if err != nil {
 			log.Logger().Warnf("could not find git URL in dir %s due to: %s", dir, err.Error())
 		} else {
-			importURL = gitInfo.HttpCloneURL()
+			gitKind, _ := o.GitServerKind(gitInfo)
```

Comment:
That really looks quite silly. So you pass _gitInfo_ into something to get the _gitKind_ which you then pass into _gitInfo_ to get the clone URL. To me, that shows that something is not quite right here. 

I feel `gitInfo.HttpCloneURL(gitKind)` is really a bad idea. I would expect `GitRepository` to be exactly the thing which hides/abstracts this from me. 

So what's about for example if we back the kind into the (various) "constructor" functions of `GitRepository`, in this case, `o.FindGitInfo(dir)`. It could be `o.GitInfoDir(dir, kind)`. 

BTW, there is nothing in the git config dir which would indicate/help us to determine whether we are dealing with BitBucket, is it? Also, if the directory got cloned via the HTTP URL to start with, we are able to even already handle this when we parse the clone URL from the git config, right? The issue would be if the clone URL in the git config is ssh based and we want to convert it to a `HttpCloneURL`. Is that correct?

BTW, is this 'scm' path segment BitBucket version dependent? Assuming BitBucket lies behind this URL - https://bitbucket.org/m2m/cumulocity-examples/src/develop/ - I am wondering why there is no 'scm' there? I am also wondering whether one can generally say "if bitbucket, then add 'scm'"?

Reply:
Yeah, there’s nothing in the git config. It’s fairly infuriating. BitBucket Cloud (I.e., bitbucket.org) doesn’t have the `scm` bit, but BitBucket Server always does.

The reason for the silliness of passing `gitInfo` to get the kind and then passing the kind to `gitInfo` is me trying to do minimal changes and leveraging the existing https://github.com/jenkins-x/jx/blob/master/pkg/cmd/opts/git.go#L174 in command options. That said, I still had to hack around that in one case, so lemme rethink this a bit.

Reply:
Blech, I can't come up with a consistent way to pull this off other than this. In _most_ cases, we're going to be getting the right `https` clone URL from the git config, but not always - someone could clone with `git@` etc, at which point we're hosed if we don't do this. We'd still need to generate an http clone URL in those cases without any way of knowing that it needs the `scm`. We also don't want to be checking against the config map/secrets/etc every time we parse a git URL, so I don't think we can have `ParseGitURL` set the kind.

Sooooo I think this is what we're stuck with.

Reply:
> BitBucket Cloud (I.e., bitbucket.org) doesn’t have the scm bit, but BitBucket Server always does.

And we can differentiate this? `KindBitBucketServer` != BitBucket Cloud?

Reply:
Yup, `KindBitBucketServer` and `KindBitBucketCloud` are different things.

Reply:
> We also don't want to be checking against the config map/secrets/etc every time we parse a git URL, so I don't think we can have ParseGitURL set the kind.

But in most cases I see, the parking of the URL is just a few lines above the line you use to determine the kind, right?
We could have multiple ParseURL methods w/ amd w/o the kind. Internally we store the kind and handle it if it is set otherways we use some sort of default. 
Personally, I have a bad feeling about this. I would poke at this a bit more, but if you think it should be merged, let's do it.

Reply:
> I'm definitely open to alternatives, I just haven't been able to find something that's consistent and correct.

Here is another idea. See whether you like this. The rub is really imo that the separation of concerns is really bad here. What if one were to remove `HttpCloneURL` from `GitRepository` (only called 5 times really as we also see here in this PR). Then we make GitKind a proper type (which is in itself probably a good thing). It still could be a simple string, but making it a type we can add methods on it. We can then move `HttpCloneURL` to `GitKind` passing the GitInfo itself or the parts which are needed to construct the URL. That might be a better flow. WDYT?

Reply:
Hmmm...I'm not sure we really want to switch git kind to a real type - seems like that may be more complexity than is worth adding for this. But moving `HttpCloneURL` to a standalone function taking a `GitRepository` and a git kind as arguments does seem more than reasonable.

Reply:
> But moving HttpCloneURL to a standalone function taking a GitRepository and a git kind as arguments does seem more than reasonable.

Sounds like a possibility as well

Reply:
And done.